File: 2024/day19/go/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// countWays returns the number of ways to construct design from patterns.
// Uses bottom-up dynamic programming with a slice for O(1) memo access.
func countWays(design string, patterns []string) int64 {
	n := len(design)
	dp := make([]int64, n+1)
	dp[n] = 1 // Base case: empty suffix has one way

	for pos := n - 1; pos >= 0; pos-- {
		for _, pattern := range patterns {
			plen := len(pattern)
			if pos+plen <= n && design[pos:pos+plen] == pattern {
				dp[pos] += dp[pos+plen]
			}
		}
	}
	return dp[0]
}

// solve computes both part1 and part2 in a single pass over designs.
// Returns the count of possible designs and total number of arrangements.
func solve(patterns, designs []string) (int, int64) {
	var possible int
	var totalWays int64
	for _, design := range designs {
		ways := countWays(design, patterns)
		if ways > 0 {
			possible++
		}
		totalWays += ways
	}
	return possible, totalWays
}

func main() {
	exePath, err := os.Executable()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting executable path: %v\n", err)
		os.Exit(1)
	}
	exeDir := filepath.Dir(exePath)
	inputPath := filepath.Join(exeDir, "..", "input.txt")

	// Fallback for running with go run
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		inputPath = filepath.Join(filepath.Dir(os.Args[0]), "..", "input.txt")
		if _, err := os.Stat(inputPath); os.IsNotExist(err) {
			// Try relative to current working directory
			inputPath = "../input.txt"
		}
	}

	data, err := os.ReadFile(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	content := strings.TrimSpace(string(data))
	parts := strings.Split(content, "\n\n")

	// Parse patterns
	patternParts := strings.Split(parts[0], ",")
	patterns := make([]string, len(patternParts))
	for i, p := range patternParts {
		patterns[i] = strings.TrimSpace(p)
	}

	// Parse designs
	designs := strings.Split(strings.TrimSpace(parts[1]), "\n")

	part1, part2 := solve(patterns, designs)
	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}

File: 2024/day19/rust/solution.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn count_ways(design: &[u8], patterns: &[&[u8]]) -> u64 {
    let mut memo: HashMap<usize, u64> = HashMap::new();
    dp(0, design, patterns, &mut memo)
}

fn dp(pos: usize, design: &[u8], patterns: &[&[u8]], memo: &mut HashMap<usize, u64>) -> u64 {
    if pos == design.len() {
        return 1;
    }

    if let Some(&cached) = memo.get(&pos) {
        return cached;
    }

    let total: u64 = patterns
        .iter()
        .filter(|p| design[pos..].starts_with(p))
        .map(|p| dp(pos + p.len(), design, patterns, memo))
        .sum();

    memo.insert(pos, total);
    total
}

fn part1(patterns: &[&[u8]], designs: &[&[u8]]) -> usize {
    designs
        .iter()
        .filter(|d| count_ways(d, patterns) > 0)
        .count()
}

fn part2(patterns: &[&[u8]], designs: &[&[u8]]) -> u64 {
    designs.iter().map(|d| count_ways(d, patterns)).sum()
}

fn main() {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let input = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| fs::read_to_string("../input.txt").expect("Failed to read input.txt"));
    let input = input.trim();

    let parts: Vec<&str> = input.split("\n\n").collect();
    let patterns: Vec<&[u8]> = parts[0].split(',').map(|s| s.trim().as_bytes()).collect();
    let designs: Vec<&[u8]> = parts[1].lines().map(|s| s.as_bytes()).collect();

    println!("Part 1: {}", part1(&patterns, &designs));
    println!("Part 2: {}", part2(&patterns, &designs));
}
